ReactWindows/ReactNative.Shared/ReactInstanceManager.cs

```diff
@@ -610,14 +612,23 @@ private void InvokeDefaultOnBackPressed()
             reactInstance.GetJavaScriptModule<AppRegistry>().runApplication(jsAppModuleName, appParameters);
         }
 
-        private void DetachViewFromInstance(ReactRootView rootView, IReactInstance reactInstance)
+        private async Task DetachViewFromInstanceAsync(ReactRootView rootView, IReactInstance reactInstance)
         {
             DispatcherHelpers.AssertOnDispatcher();
 
             var uiManagerModule = reactInstance.GetNativeModule<UIManagerModule>();
-            uiManagerModule.DetachRootView(rootView);
+
+            // Detaches ReactRootView from instance manager root view list and size change monitoring.
+            // Has to complete before unmounting application
+            await uiManagerModule.DetachRootViewAsync(rootView);
+
+            // Initiates a clean up operation that will wait for both a "removeRootView" command from JS (an effect of
+            // unmounting the application) and the release of all dispatcher affined UI objects.
+            var crTask = uiManagerModule.GetRootViewCleanupTask(rootView.GetTag());
 
```

Comment:
Any reason not to embed this logic in DetachRootViewAsync?

Reply:
I would've had to return "crTask" separately, somehow, as an out param maybe.
We have to call unmountApplication "before" awaiting on crTask, otherwise you deadlock.

Looked cleaner this way.

---
In reply to: [179482244](https://github.com/Microsoft/react-native-windows/pull/1737#discussion_r179482244) [](ancestors = 179482244)

Reply:
We could add another method like:
`WaitForUnmountApplicationAsync`

It might be a bit more explicit for what we're doing.

---
In reply to: [179483720](https://github.com/Microsoft/react-native-windows/pull/1737#discussion_r179483720) [](ancestors = 179483720,179482244)

Reply:
So:
```
            // Detaches ReactRootView from instance manager root view list and size change monitoring.
            // Has to complete before unmounting application
            await uiManagerModule.DetachRootViewAsync(rootView);

            reactInstance.GetJavaScriptModule<AppRegistry>().unmountApplicationComponentAtRootTag(rootView.GetTag());

            await uiManagerModule.WaitForUnmountApplicationAsync(rootView.GetTag());

---
In reply to: [179488632](https://github.com/Microsoft/react-native-windows/pull/1737#discussion_r179488632) [](ancestors = 179488632,179483720,179482244)

Reply:
Or, for that matter, encapsulate all this logic in a single helper method on UIManagerModule. Something like:
```
UIManagerModule.UnmountApplicationAsync(rootView);
```

You could get ahold of the `AppRegistry` JavaScript module from the `ReactContext` inside `UIManagerModule`

---
In reply to: [179488909](https://github.com/Microsoft/react-native-windows/pull/1737#discussion_r179488909) [](ancestors = 179488909,179488632,179483720,179482244)

Reply:
Tough call.

The second option (one method) can do things right, but it would break the symmetry we have with AttachMeasuredRootViewToInstance (that mounts the application)
The first option may look better, but it can't implement the right sequencing: to keep the race condition windows totally shut you have to create the TaskCompletionSource based task before the actual unmounting of the application...

So.. which is it?



---
In reply to: [179493061](https://github.com/Microsoft/react-native-windows/pull/1737#discussion_r179493061) [](ancestors = 179493061,179488909,179488632,179483720,179482244)

Reply:
The first suggestion was to create the TaskCompletionSource as part of the `DetachRootViewAsync` method. It seems that `unmountApplication...` always follows `DetachRootView`. I'd lean towards keeping the surface area small. We can also create symmetry in the first one by calling `startApplication` from UIManagerModule as well.

Reply:
Ok, I’ll see how it goes and what time I have, I’m partying right now.